File: src/main/java/ca/yyx/hu/fragments/NetworkListFragment.java
package ca.yyx.hu.fragments;

import android.app.DialogFragment;
import android.app.FragmentManager;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Html;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import java.io.IOException;
import java.net.InetAddress;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;

import ca.yyx.hu.R;
import ca.yyx.hu.aap.AapService;
import ca.yyx.hu.utils.NetworkUtils;
import ca.yyx.hu.utils.Settings;

/**
 * @author algavris
 * @date 05/11/2016.
 */

public class NetworkListFragment extends BaseFragment {
    public static final String TAG = "NetworkListFragment";
    AddressAdapter mAdapter;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        RecyclerView recyclerView = (RecyclerView) inflater.inflate(R.layout.fragment_list, container, false);

        Context context = getContext();

        mAdapter = new AddressAdapter(context, getFragmentManager());
        recyclerView.setLayoutManager(new LinearLayoutManager(context));
        recyclerView.setAdapter(mAdapter);
        return recyclerView;
    }

    @Override
    public void onResume() {
        super.onResume();

        try {
            int currentIp = NetworkUtils.getWifiIpAddress(getActivity());
            InetAddress inet = NetworkUtils.intToInetAddress(currentIp);
            mAdapter.setCurrentAddress(inet);
        } catch (IOException ignored) {
            mAdapter.setNoCurrentAddress();
        }

        mAdapter.loadAddresses();
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    public void addAddress(InetAddress ip) {
        mAdapter.addNewAddress(ip);
    }

    private static class DeviceViewHolder extends RecyclerView.ViewHolder {
        final Button allowButton;
        final Button startButton;

        DeviceViewHolder(View itemView) {
            super(itemView);
            this.startButton = (Button) itemView.findViewById(android.R.id.button2);
            this.allowButton = (Button) itemView.findViewById(android.R.id.button1);
        }
    }

    private static class AddressAdapter extends RecyclerView.Adapter<DeviceViewHolder> implements View.OnClickListener {
        private final Context mContext;
        private final FragmentManager mFragmentManager;
        private ArrayList<String> mAddressList = new ArrayList<>();
        private InetAddress mCurrentAddress;
        private final Settings mSettings;


        AddressAdapter(Context context, FragmentManager fragmentManager) {
            mContext = context;
            mFragmentManager = fragmentManager;
            mSettings = new Settings(context);

        }

        @Override
        public DeviceViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(mContext).inflate(R.layout.list_item_device, parent, false);
            DeviceViewHolder holder = new DeviceViewHolder(view);

            holder.startButton.setOnClickListener(this);
            holder.allowButton.setOnClickListener(this);
            return holder;
        }

        @Override
        public void onBindViewHolder(DeviceViewHolder holder, int position) {
            String device = mAddressList.get(position);

            String line1;
            String line2;
            if (position == 0) {
                line1 = "Add a new address";
                line2 = "Current ip: " + (TextUtils.isEmpty(device) ? "No ip address" : device);
                holder.allowButton.setVisibility(View.GONE);
            } else {
                line1 = device;
                line2 = "";
                holder.allowButton.setVisibility(View.VISIBLE);
                holder.allowButton.setText(R.string.remove);
            }
            String msg = String.format(Locale.US, "<b>%1$s</b><br/>%2$s", line1, line2);
            holder.startButton.setTag(R.integer.key_position, position);
            holder.startButton.setText(Html.fromHtml(msg));
            holder.startButton.setTag(R.integer.key_data, device);
            holder.allowButton.setText(R.string.remove);
        }

        @Override
        public int getItemCount() {
            return mAddressList.size();
        }

        @Override
        public void onClick(View v) {
            if (v.getId() == android.R.id.button2) {
                int position = (int) v.getTag(R.integer.key_position);
                if (position == 0) {
                    InetAddress ip = null;
                    try {
                        int ipInt = NetworkUtils.getWifiIpAddress(mContext);
                        ip = NetworkUtils.intToInetAddress(ipInt);
                    } catch (IOException ignored) {
                    }

                    AddNetworkAddressDialog dialog = AddNetworkAddressDialog.create(ip);
                    dialog.show(mFragmentManager, "AddNetworkAddressDialog");
                } else {
                    String ipAddress = (String) v.getTag(R.integer.key_data);
                    mContext.startService(AapService.createIntent(ipAddress, mContext));
                }
            } else {
                String ipAddress = (String) v.getTag(R.integer.key_data);
                this.removeAddress(ipAddress);
            }
        }

        private void addCurrentAddress() {
            if (mCurrentAddress != null) {
                mAddressList.add(mCurrentAddress.getHostAddress());
            } else {
                mAddressList.add("");
            }
        }

        void setCurrentAddress(InetAddress currentAddress) {
            mCurrentAddress = currentAddress;
        }

        void setNoCurrentAddress() {
            mCurrentAddress = null;
        }

        void addNewAddress(InetAddress ip) {
            HashSet<String> addrs = (HashSet<String>) mSettings.getNetworkAddresses();
            addrs.add(ip.getHostAddress());
            mSettings.setNetworkAddresses(addrs);
            set(addrs);
        }

        void loadAddresses() {
            Set<String> addrs = mSettings.getNetworkAddresses();
            set(addrs);
        }

        private void set(Collection<String> addrs) {
            mAddressList.clear();
            addCurrentAddress();
            mAddressList.addAll(addrs);
            notifyDataSetChanged();
        }

        private void removeAddress(String ipAddress) {
            HashSet<String> addrs = (HashSet<String>) mSettings.getNetworkAddresses();
            addrs.remove(ipAddress);
            mSettings.setNetworkAddresses(addrs);
            set(addrs);
        }

    }
}

File: src/main/java/ca/yyx/hu/main/NetworkListFragment.kt
package ca.yyx.hu.main

import android.app.FragmentManager
import android.content.Context
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.Html
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button

import java.io.IOException
import java.net.InetAddress
import java.util.ArrayList
import java.util.HashSet
import java.util.Locale

import ca.yyx.hu.R
import ca.yyx.hu.aap.AapService
import ca.yyx.hu.app.BaseFragment
import ca.yyx.hu.utils.NetworkUtils
import ca.yyx.hu.utils.Settings

/**
 * @author algavris
 * *
 * @date 05/11/2016.
 */

class NetworkListFragment : BaseFragment() {
    private lateinit var mAdapter: AddressAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup, savedInstanceState: Bundle?): View {
        val recyclerView = inflater.inflate(R.layout.fragment_list, container, false) as RecyclerView

        val context = activity

        mAdapter = AddressAdapter(context, fragmentManager)
        recyclerView.layoutManager = LinearLayoutManager(context)
        recyclerView.adapter = mAdapter
        return recyclerView
    }

    override fun onResume() {
        super.onResume()

        try {
            val currentIp = NetworkUtils.getWifiIpAddress(activity)
            val inet = NetworkUtils.intToInetAddress(currentIp)
            mAdapter.setCurrentAddress(inet)
        } catch (ignored: IOException) {
            mAdapter.setNoCurrentAddress()
        }

        mAdapter.loadAddresses()
    }

    fun addAddress(ip: InetAddress) {
        mAdapter.addNewAddress(ip)
    }

    private class DeviceViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        internal val allowButton: Button = itemView.findViewById(android.R.id.button1) as Button
        internal val startButton: Button = itemView.findViewById(android.R.id.button2) as Button
    }

    private class AddressAdapter
        internal constructor(
                private val mContext: Context,
                private val mFragmentManager: FragmentManager) : RecyclerView.Adapter<DeviceViewHolder>(), View.OnClickListener {

        private val mAddressList = ArrayList<String>()
        private var mCurrentAddress: InetAddress? = null
        private val mSettings: Settings = Settings(mContext)

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DeviceViewHolder {
            val view = LayoutInflater.from(mContext).inflate(R.layout.list_item_device, parent, false)
            val holder = DeviceViewHolder(view)

            holder.startButton.setOnClickListener(this)
            holder.allowButton.setOnClickListener(this)
            return holder
        }

        override fun onBindViewHolder(holder: DeviceViewHolder, position: Int) {
            val device = mAddressList[position]

            val line1: String
            val line2: String
            if (position == 0) {
                line1 = "Add a new address"
                line2 = "Current ip: " + if (TextUtils.isEmpty(device)) "No ip address" else device
                holder.allowButton.visibility = View.GONE
            } else {
                line1 = device
                line2 = ""
                holder.allowButton.visibility = View.VISIBLE
                holder.allowButton.setText(R.string.remove)
            }
            val msg = String.format(Locale.US, "<b>%1\$s</b><br/>%2\$s", line1, line2)
            holder.startButton.setTag(R.integer.key_position, position)
            holder.startButton.text = Html.fromHtml(msg)
            holder.startButton.setTag(R.integer.key_data, device)
            holder.allowButton.setText(R.string.remove)
        }

        override fun getItemCount(): Int {
           return mAddressList.size
        }

        override fun onClick(v: View) {
            if (v.id == android.R.id.button2) {
                if (v.getTag(R.integer.key_position) == 0) {
                    var ip: InetAddress? = null
                    try {
                        val ipInt = NetworkUtils.getWifiIpAddress(mContext)
                        ip = NetworkUtils.intToInetAddress(ipInt)
                    } catch (ignored: IOException) {
                    }

                    val dialog = AddNetworkAddressDialog.create(ip)
                    dialog.show(mFragmentManager, "AddNetworkAddressDialog")
                } else {
                    mContext.startService(AapService.createIntent(v.getTag(R.integer.key_data) as String, mContext))
                }
            } else {
                this.removeAddress(v.getTag(R.integer.key_data) as String)
            }
        }

        private fun addCurrentAddress() {
            if (mCurrentAddress != null) {
                mAddressList.add(mCurrentAddress!!.getHostAddress())
            } else {
                mAddressList.add("")
            }
        }

        internal fun setCurrentAddress(currentAddress: InetAddress) {
            mCurrentAddress = currentAddress
        }

        internal fun setNoCurrentAddress() {
            mCurrentAddress = null
        }

        internal fun addNewAddress(ip: InetAddress) {
            val addrs = mSettings.networkAddresses as HashSet<String>
            addrs.add(ip.hostAddress)
            mSettings.networkAddresses = addrs
            set(addrs)
        }

        internal fun loadAddresses() {
            val addrs = mSettings.networkAddresses
            set(addrs)
        }

        private fun set(addrs: Collection<String>) {
            mAddressList.clear()
            addCurrentAddress()
            mAddressList.addAll(addrs)
            notifyDataSetChanged()
        }

        private fun removeAddress(ipAddress: String) {
            val addrs = mSettings.networkAddresses as HashSet<String>
            addrs.remove(ipAddress)
            mSettings.networkAddresses = addrs
            set(addrs)
        }

    }

    companion object {
        val TAG = "NetworkListFragment"
    }
}
